src/transformers/models/opt/modeling_tf_opt.py

```diff
@@ -0,0 +1,1085 @@
+        all_self_attns = () if output_attentions else None
+        present_key_values = () if use_cache else None
+
+        # check if head_mask and cross_attn_head_mask have a correct number of layers specified if desired
+        # The tf.debugging asserts are not compliant with XLA then they
+        # have to be disabled in other modes than eager.
+        for attn_mask_name, attn_mask in [("head_mask", head_mask)]:
+            if attn_mask is not None and tf.executing_eagerly():
+                tf.debugging.assert_equal(
+                    shape_list(attn_mask)[0],
+                    len(self.layers),
+                    message=(
+                        f"The {attn_mask_name} should be specified for {len(self.layers)} layers, but it is for"
+                        f" {shape_list(attn_mask)[0]}."
+                    ),
+                )
+
+        for idx, decoder_layer in enumerate(self.layers):
+            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+
+            # dropout_probability = random.uniform(0, 1)
+
+            # if training and (dropout_probability < self.layerdrop):
+            #     continue
+
+            past_key_value = past_key_values[idx] if past_key_values is not None else None
+
+            hidden_states, layer_self_attn, present_key_value = decoder_layer(
+                hidden_states,
+                attention_mask=attention_mask,
+                layer_head_mask=head_mask[idx] if head_mask is not None else None,
+                past_key_value=past_key_value,
+            )
+
+            if use_cache:
+                present_key_values += (present_key_value,)
+
+            if output_attentions:
+                all_self_attns += (layer_self_attn,)
+
+        if self.project_out is not None:
+            hidden_states = self.project_out(hidden_states)
+
+        if output_hidden_states:
+            all_hidden_states += (hidden_states,)
+
+        if not return_dict:
+            return tuple(
+                v for v in [hidden_states, present_key_values, all_hidden_states, all_self_attns] if v is not None
+            )
+
+        else:
+            return TFBaseModelOutputWithPast(
+                last_hidden_state=hidden_states,
+                past_key_values=present_key_values,
+                hidden_states=all_hidden_states,
+                attentions=all_self_attns,
+            )
+
+
+@add_start_docstrings(
+    "The bare TF OPT Model outputting raw hidden-states without any specific head on top.",
+    OPT_START_DOCSTRING,
+)
+@keras_serializable
+class TFOPTModel(TFPreTrainedModel):
+    config_class = OPTConfig
+
+    def __init__(self, config: OPTConfig, **kwargs):
+        super().__init__(config, **kwargs)
+        self.config = config
+        self.decoder = TFOPTMainLayer(config, name="decoder")
```

Comment:
why decoder ? think it should be called `"model"`

Reply:
I guess it was to follow the pytorch implementation where the `OPTModel` has a `decoder` attribute. Since the `TFOPTMainLayer` follows the `OPTDecoder` implementation. WDYT? It makes more sens for me to keep the `model.decoder`, changing would give

```python
>>> model = TFOPTModel.from_pretrained("facebook/opt-350m")
>>> model.summary()
Model: "tfopt_model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 model (TFOPTMainLayer)    multiple                  331196416 
                                                                 
=================================================================
Total params: 331,196,416
Trainable params: 331,196,416
Non-trainable params: 0
_________________________________________________________________
```
And we would acces the decoder through `model.model`  which does not really make sense for me.

Reply:
Agreed, I was a bit misslead since the `MainLayer` corresponds to the `OPTDecoder` in torch. Thanks

Reply:
The important part here is that the automatic conversion between PyTorch and Tensorflow works correctly and that it's aligned with PyTorch. 

I don't really see why we would do `"decoder"` is better than `"model"` here. 
E.g. you would do:

```py
opt = TFOPTModel.from_pretrained(....)
opt.model = <to/access/main/layer>
```
no? 

We should not change weight names as it would break the automatic conversion which is **very** important IMO.